File: 20230313_Monday/20230313.rb
# Leetcode: 101. Symmetric Tree.
# https://leetcode.com/problems/symmetric-tree/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 83 ms, faster than 76.47% of Ruby online submissions for Symmetric Tree.
# Memory Usage: 211 MB, less than 43.53% of Ruby online submissions for Symmetric Tree.
# 2023.03.13 Daily Challenge.
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_symmetric(root)
  q = [root]
  while ! q.empty?
    nq = [] # new queue.
    return false if ! sym(q)
    q.each do |x|
      next if ! x
      nq.push(x.left)
      nq.push(x.right)
    end
    q = nq
  end
  return true
end

def sym(arr)
  i,j = 0, arr.size - 1
  while i < j
    if arr[i].nil? || arr[j].nil?
      return false if arr[i] || arr[j]
    else
      return false if arr[i].val != arr[j].val
    end
    # Need to move pointers on each interation.
    i += 1
    j -= 1
  end
  return true
end

File: 20230313_Monday/20230313.go
// Leetcode: 101. Symmetric Tree.
// https://leetcode.com/problems/symmetric-tree/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 2 ms, faster than 66.57% of Go online submissions for Symmetric Tree.
// Memory Usage: 3 MB, less than 14.72% of Go online submissions for Symmetric Tree.
// 2023.03.14 Updated.
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
	q := make([]*TreeNode, 0)
	nq := make([]*TreeNode, 0)
	q = append(q, root)
	for 0 != len(q) {
		nq = nq[:0]
		if !sym(q) {
			return false
		}
		for _, x := range q {
			if nil == x {
				continue
			}
			nq = append(nq, x.Left)
			nq = append(nq, x.Right)
		}
		q, nq = nq, q
	}
	return true
}

func sym(arr []*TreeNode) bool {
	i, j := 0, len(arr)-1
	for i < j {
		if nil == arr[i] || nil == arr[j] {
			if nil != arr[i] || nil != arr[j] {
				return false
			}
		} else {
			if arr[i].Val != arr[j].Val {
				return false
			}
		}
		i += 1
		j -= 1
	}
	return true
}
